### PR TITLE
[RFC] ASoC: SOF: merge ipc and stream irq

### PR DESCRIPTION
**Pull request by bardliao**
We may lost interrupt on MSI mode if two interrupts are triggered at
the same time. Use single handler and check both interrupts can avoid
that.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

### CODE REVIEW

File: sound/soc/sof/intel/hda.c
Comment (RanderWang) [inline sound/soc/sof/intel/hda.c]: how about to change hda_dsp_stream_interrupt to hda_dsp_stream_irq_handler like hda_dsp_ipc_irq_handler

Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: I would change the return type of those functions to a boolean, e.g.

````c
if (hda_dsp_check_ipc_irq(irq, context) || 
    hda_dsp_check_stream_irq(irq, context))
        return IRQ_WAKE_THREAD;
````

File: sound/soc/sof/intel/hda-ipc.c
Comment (plbossart) [inline sound/soc/sof/intel/hda-ipc.c:247]: question:
I can never remember the rules for spin_locks.

Do we need to protect with the _irq suffix only here, or also in the thread?

File: sound/soc/sof/intel/hda-ipc.c
Comment (plbossart) [inline sound/soc/sof/intel/hda-ipc.c]: maybe make the commit message more explicit:

ASoC: SOF: Intel: hda: fix MSI issues by merging ipc and stream irq handlers

The existing code uses two handlers for a shared edge-based MSI interrupts. In corner cases,  interrupts are lost, leading to IPC timeouts. Those timeouts do not appear in legacy mode.

This patch merges the two handlers into a single one. Follow-up patches will extend this work to SoundWire interrupts which are also shared with IPC and stream interrupts.

**Review by plbossart — CHANGES_REQUESTED** [review]
Thanks @bardliao, looks mostly good, I suggested some improvements to make the code more readable.

File: sound/soc/sof/intel/hda-ipc.c
Comment (paulstelian97) [reply sound/soc/sof/intel/hda-ipc.c:247]: I believe it has to be protected with _irq in the thread and not necessarily in the IRQ handler (you'd use _irq if you're not 100% certain the code runs only in IRQ handlers). IIRC, that is.

Comment (bardliao) [reply sound/soc/sof/intel/hda-ipc.c:247]: Thanks @paulstelian97 for the explanation . @plbossart hda_dsp_ipc_irq_handler() will be called in the irq thread that's why I add the _irq suffix here. But actually I don't know if we do need the spin_lock or not. @RanderWang discussed it with me about it, but we didn't have a conclusion.

File: sound/soc/sof/intel/hda.c
Comment (bardliao) [reply sound/soc/sof/intel/hda.c]: Thanks @RanderWang . I renamed them to is_stream_irq() and is_ipc_irq()

Comment (bardliao) [reply sound/soc/sof/intel/hda.c]: @plbossart Done. But I add a new function instead of changing  hda_dsp_ipc_irq_handler() since it is used in somewhere else.

**Comment by bardliao** [comment]
I use work in hda_dsp_interrupt_threaded() since we want to handle different interrupt in parallel.  And it can prevent one is blocked by the other.

File: sound/soc/sof/intel/hda-stream.c
Comment (plbossart) [inline sound/soc/sof/intel/hda-stream.c]: bool

File: sound/soc/sof/intel/hda-ipc.c
Comment (plbossart) [inline sound/soc/sof/intel/hda-ipc.c]: can't you remove this function and do all the work here?
Why is a wrapper needed with a first parameter 0 that seems useless?

File: sound/soc/sof/intel/hda.c
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: this is really odd, you are not using the first argument 0?

File: sound/soc/sof/intel/hda.c
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: is this meant to handle the APL/CNL difference only?

**Review by plbossart — CHANGES_REQUESTED** [review]

File: sound/soc/sof/intel/hda.c
Comment (bardliao) [reply sound/soc/sof/intel/hda.c]: No, the first argument is never used.

File: sound/soc/sof/intel/hda.c
Comment (bardliao) [reply sound/soc/sof/intel/hda.c]: Yes, currently, hda_dsp_probe() is for APL/CNL only and they have different .irq_thread.

File: sound/soc/sof/intel/hda-ipc.c
Comment (bardliao) [reply sound/soc/sof/intel/hda-ipc.c]: Oh, I just want to reuse the code. Yes, I can remove that function.

File: sound/soc/sof/intel/hda.c
Comment (RanderWang) [inline sound/soc/sof/intel/hda.c]: @bardliao  why we need to process interrupt in second level scheduled work ?   hda_dsp_interrupt_threaded is in a work scheduled by interrupt framework.

Comment (bardliao) [reply sound/soc/sof/intel/hda.c]: @RanderWang hda_dsp_stream_threaded_handler() could take a long time, and we don't wan't one threaded handler to wait for the other. Besides, we may want to use loop on each threaded handler in the future.

Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: @RanderWang has a point though.what would happen if we didn't not use a threaded handler, and called schedule_work directly from the handler?

File: sound/soc/sof/intel/hda-stream.c
Comment (kv2019i) [inline sound/soc/sof/intel/hda-stream.c:569]: This is a bit asymmetric in that is_ipc() function disabled IPC interrupts in this check, while this function does not disable the interrupt source? Why is this?

File: sound/soc/sof/intel/hda.c
Comment (kv2019i) [inline sound/soc/sof/intel/hda.c:548]: @bardliao Shouldn't this patch remove hda_dsp_ipc_irq_handler as well?

File: sound/soc/sof/intel/hda.c
Comment (kv2019i) [reply sound/soc/sof/intel/hda.c]: @bardliao I'm a bit concerned about this as well. We are essentially adding two mandatory context switches for every interrupt. First from hard-irq to irq thread, and then from irq thread to work-queue. Do you have concrete logs showing that the threaded handler is taking a lot of time?

**Review by kv2019i — CHANGES_REQUESTED** [review]
As overall design, I think this is the way to go. I have some concerns on the specific implementation, see comments inline.

File: sound/soc/sof/intel/hda-ipc.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-ipc.c:247]: @bardliao I'm not sure about this. I dont think we need to use spin_lock_irq(). Back when we're still upstreaming the SOF driver, IIRC the comment upstream was that spin_lock_irq() should only be in the hard interrupt handler.

File: sound/soc/sof/intel/hda.c
Comment (ranj063) [reply sound/soc/sof/intel/hda.c]: I also agree that this seems a bit of an overkill. Instead we should just wake the appropriate thread in the handler based on is_ipc_irq() or is_stream_irq() is true.

File: sound/soc/sof/intel/hda.c
Comment (ranj063) [inline sound/soc/sof/intel/hda.c:531]: @bardliao if this is a common irq, we should consider renaming it to maybe just irq?

File: sound/soc/sof/intel/hda.c
Comment (ranj063) [reply sound/soc/sof/intel/hda.c]: @bardliao a general comment about the naming of the functions. We should only call the hard interrupt handler as somthing_thread_handler(), the thread itself should just be called irq_thread or the stream_thread

**Comment by ranj063** [comment]
@bardliao there was specific change we made in the stream interrupt handler to handled unsolicited responses from the HDA codecs (ie looping 10 times) and checking INTSTS. This is was fixed the IPC timeouts for us. Now, in this PR, you are suggesting moving it to the IRQ thread. I'm a bit concerned this could cause regressions though.

**Comment by plbossart** [comment]
> @bardliao there was specific change we made in the stream interrupt handler to handled unsolicited responses from the HDA codecs (ie looping 10 times) and checking INTSTS. This is was fixed the IPC timeouts for us. Now, in this PR, you are suggesting moving it to the IRQ thread. I'm a bit concerned this could cause regressions though.

that was already done in the interrupt thread, but now it'd be moved to a workqueue and indeed this could impact the results

Since we are moving to a single thread, I wonder if the simple thing to do is to sequentially deal with IPC, then stream, then SoundWire (without using schedule_work as suggested so far). In the latter case, we actually use a work-queue already so there wouldn't be any change performance-wise. Also the peak times for interrupts don't overlap. For the firmware setup, we have a slew of IPC, but they happen before SoundWire kicks in. then we have stream interrupts while the flows are setup. It's not completely random, the control parts peak before the streaming.

The only change would come from the serialization of IPC and stream interrupts, but that doesn't look so bad to me.

**Comment by ranj063** [comment]
> that was already done in the interrupt thread, but now it'd be moved to a workqueue and indeed this could impact the results

@plbossart I am talking about this part: https://github.com/thesofproject/linux/blob/ecc776aef3085e3f45a6b8d4e22c7a5cd5fd85cc/sound/soc/sof/intel/hda-stream.c#L617

This is still done in the handler and not in the thread. But with @bardliao 's changes this would be moved to the workqueue or if he changes it to the interrupt thread, either way I feel like this could cause a regression

**Comment by plbossart** [comment]
> > that was already done in the interrupt thread, but now it'd be moved to a workqueue and indeed this could impact the results
> 
> @plbossart I am talking about this part:
> 
> https://github.com/thesofproject/linux/blob/ecc776aef3085e3f45a6b8d4e22c7a5cd5fd85cc/sound/soc/sof/intel/hda-stream.c#L617
> 
> This is still done in the handler and not in the thread. But with @bardliao 's changes this would be moved to the workqueue or if he changes it to the interrupt thread, either way I feel like this could cause a regression

nope, the handler is here:
https://github.com/thesofproject/linux/blob/ecc776aef3085e3f45a6b8d4e22c7a5cd5fd85cc/sound/soc/sof/intel/hda-stream.c#L552

**Comment by ranj063** [comment]
> > > that was already done in the interrupt thread, but now it'd be moved to a workqueue and indeed this could impact the results
> > 
> > 
> > @plbossart I am talking about this part:
> > https://github.com/thesofproject/linux/blob/ecc776aef3085e3f45a6b8d4e22c7a5cd5fd85cc/sound/soc/sof/intel/hda-stream.c#L617
> > 
> > This is still done in the handler and not in the thread. But with @bardliao 's changes this would be moved to the workqueue or if he changes it to the interrupt thread, either way I feel like this could cause a regression
> 
> nope, the handler is here:
> https://github.com/thesofproject/linux/blob/ecc776aef3085e3f45a6b8d4e22c7a5cd5fd85cc/sound/soc/sof/intel/hda-stream.c#L552

duh! yes you are correct @plbossart. the name threw me off a bit. In this case, if we dont use the workqueue, we should be fine I think

File: sound/soc/sof/intel/hda.c
Comment (bardliao) [reply sound/soc/sof/intel/hda.c]: Thank you all. Yes, it would be better to wake the appropriate thread in the handler.

File: sound/soc/sof/intel/hda-ipc.c
Comment (bardliao) [reply sound/soc/sof/intel/hda-ipc.c:247]: Thanks. I will double check it.

File: sound/soc/sof/intel/hda-stream.c
Comment (bardliao) [reply sound/soc/sof/intel/hda-stream.c:569]: Good question. I think we should disable it as well, but we can do it in another PR.

File: sound/soc/sof/intel/hda.c
Comment (bardliao) [reply sound/soc/sof/intel/hda.c:531]: Yeah, I also think about it. However, for those platforms without HDA, the irq is still for ipc only. That's why I leave it as it is for now.

File: sound/soc/sof/intel/hda.c
Comment (RanderWang) [reply sound/soc/sof/intel/hda.c]: > @RanderWang hda_dsp_stream_threaded_handler() could take a long time, and we don't wan't one threaded handler to wait for the other. Besides, we may want to use loop on each threaded handler in the future.

Actually, the two works are still scheduled on the same CPU because scheduler is designed to utilize cache data (please check scheduler documents), so this can't save time and spend extra time to schedule work.

Comment (RanderWang) [reply sound/soc/sof/intel/hda.c]: > @RanderWang has a point though.what would happen if we didn't not use a threaded handler, and called schedule_work directly from the handler?

We should follow the requirement of request_threaded_irq
please check the comments for request_threaded_irq in kernel.
````
 *	If you want to set up a threaded irq handler for your device
 *	then you need to supply @handler and @thread_fn. @handler is
 *	still called in hard interrupt context and has to check
 *	whether the interrupt originates from the device. If yes it
 *	needs to disable the interrupt on the device and return
 *	IRQ_WAKE_THREAD which will wake up the handler thread and run
 *	@thread_fn. This split handler design is necessary to support
 *	shared interrupts.
````

File: sound/soc/sof/intel/hda-stream.c
Comment (RanderWang) [reply sound/soc/sof/intel/hda-stream.c:569]: yes, please check the comments for request_threaded_irq

If yes it needs to disable the interrupt on the device and return IRQ_WAKE_THREAD which will wake up the handler thread and run @thread_fn. This split handler design is necessary to support shared interrupts.

**Comment by RanderWang** [comment]
> duh! yes you are correct @plbossart. the name threw me off a bit. In this case, if we dont use the workqueue, we should be fine I think

yes, I was also confused.  handler and thread are official name for request_threaded_irq,  no **thread- handler**

File: sound/soc/sof/intel/hda.c
Comment (bardliao) [reply sound/soc/sof/intel/hda.c:548]: hda_dsp_ipc_irq_handler is assigned to cnl and apl's irq_handler ops. But no one uses it. So, yes, we can remove hda_dsp_ipc_irq_handler as well.

**Comment by bardliao** [comment]
> > duh! yes you are correct @plbossart. the name threw me off a bit. In this case, if we dont use the workqueue, we should be fine I think
> 
> yes, I was also confused. handler and thread are official name for request_threaded_irq, no **thread- handler**

@RanderWang @ranj063 We can rename it in another patch :)

**Comment by bardliao** [comment]
> > @bardliao there was specific change we made in the stream interrupt handler to handled unsolicited responses from the HDA codecs (ie looping 10 times) and checking INTSTS. This is was fixed the IPC timeouts for us. Now, in this PR, you are suggesting moving it to the IRQ thread. I'm a bit concerned this could cause regressions though.
> 
> that was already done in the interrupt thread, but now it'd be moved to a workqueue and indeed this could impact the results
> 
> Since we are moving to a single thread, I wonder if the simple thing to do is to sequentially deal with IPC, then stream, then SoundWire (without using schedule_work as suggested so far). In the latter case, we actually use a work-queue already so there wouldn't be any change performance-wise. Also the peak times for interrupts don't overlap. For the firmware setup, we have a slew of IPC, but they happen before SoundWire kicks in. then we have stream interrupts while the flows are setup. It's not completely random, the control parts peak before the streaming.
> 
> The only change would come from the serialization of IPC and stream interrupts, but that doesn't look so bad to me.

ok. Let's go with sequentially dealing with IPC, stream, and SoundWire first and see if everything works fine. :)

**Comment by RanderWang** [comment]
There is a potential risk to schedule work in thread function.  For kernel, if  thread function returns, interrupt frame work assumes all job has been done so another thread maybe invoked. But are the scheduled work done ?  Do we need to sync all the scheduled works ?   can anyone answer this question?

File: sound/soc/sof/intel/hda-ipc.c
Comment (bardliao) [reply sound/soc/sof/intel/hda-ipc.c:247]: @ranj063 I can see spin_lock() is called twice without spin_unlock() if I test it with spin_lock() instead of spin_lock_irq(). I think the reason is that the irq did not be disabled when we use spin_lock() and the irq handler could be triggered when the irq thread is still running. Also, [spin_lock_irq()](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda-stream.c#L618) is using in hda_dsp_stream_threaded_handler() which is a irq thread actually. So I believe that we should use spin_lock_irq()

Comment (kv2019i) [reply sound/soc/sof/intel/hda-ipc.c:247]: @bardliao @ranj063 You definitely need to use the "_irq()" variant here. These is_xxx_irq() functions are called both from hard-irq and softirq context, so plain spin_lock() will not work.

File: sound/soc/sof/intel/hda.c
Comment (kv2019i) [inline sound/soc/sof/intel/hda.c]: @bardliao is_ipc_irq() is called in both hardirq and softirq. This is otherwise ok, but it's a bit odd that is_ipc_irq() also has side-effects in that it clears HDA_DSP_ADSPIC_IPC. I'd prefer to call the function something else than "is_foo_irq()" as to me (maybe just ipc_irq()). To me, a function with is_something(), indicates the function does not have side-effects. The functions here do more. 

I guess this still works as the ADSPIS status bit remains high despite  ADSPIC set low. And this matches with the current logic we have.

File: sound/soc/sof/intel/hda-stream.c
Comment (kv2019i) [reply sound/soc/sof/intel/hda-stream.c:569]: Ack @bardliao and @RanderWang This is matching existing code, so I'm ok to change in another PR. I do have a comment on naming the is_ipc_irq() functions (see my other comment).

**Review by kv2019i — COMMENTED** [review]
Starts to look good.  I don't have any major concerns open. I'd like to hear your comments on the function naming, but otherwise I'm ready to press approve on this.

File: sound/soc/sof/intel/hda.c
Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: yes, this was a completely dumb comment from me. Oops :-)
I think what I was trying to say if whether the same handler can trigger multiple threads, but that seems like asking for trouble.

File: sound/soc/sof/intel/hda.c
Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: @kv2019i I had the same feeling but it's hard to move the mask outside of the spinlock, and taking the spinlock twice, once to read ADSPIC2 and once to clear ADSPIS2 is not that great either.

maybe adding a comment would help explain the hidden side effect.
or rename from is_ipc to check_ipc, which is let of a status return

Comment (bardliao) [reply sound/soc/sof/intel/hda.c]: Thanks @kv2019i @plbossart  I will rename it and add a comment.

File: sound/soc/sof/intel/hda.c
Comment (ranj063) [inline sound/soc/sof/intel/hda.c:425]: @bardliao this looks bad. You call check_ipc() in the handler, and then wake up the thread and then check_ipc() again and do the same thing over?

File: sound/soc/sof/intel/hda.c
Comment (ranj063) [reply sound/soc/sof/intel/hda.c:531]: @bardliao even in that case it makes sense to rename it to something else. Maybe in another PR

File: sound/soc/sof/intel/hda.c
Comment (bardliao) [reply sound/soc/sof/intel/hda.c:425]: @ranj063 Yes, that's odd, but there are more than one interrupt could be trigger and one may be triggered after the handler is checked. So we need to recheck it on the thread. Actually, I think we can have a flag that store the interrupt status which is checked on the handler. So we won't need to double check it on the thread again. Please see [hda_dsp_irq_thread](https://github.com/thesofproject/linux/pull/1451/files#diff-0c41ff728d2838606e1e70839d291aedR294)

File: sound/soc/sof/intel/hda.c
Comment (bardliao) [reply sound/soc/sof/intel/hda.c]: @kv2019i @plbossart done. BTW, do you know how to disable stream irq? Or we can leave it as it is for now?

File: sound/soc/sof/intel/hda.c
Comment (ranj063) [reply sound/soc/sof/intel/hda.c:425]: Why not just return IRQ_WAKE_THREAD in hda_dsp_interrupt_handler() and then handle all the checks in the thread?

Comment (bardliao) [reply sound/soc/sof/intel/hda.c:425]: > Why not just return IRQ_WAKE_THREAD in hda_dsp_interrupt_handler() and then handle all the checks in the thread?

@ranj063 We can only do it when there is not a shared irq. i.e. No one else could trigger the irq. Probably, we can do that after ipc/stream/sdw irq are merged. And we don't need the handler if it return IRQ_WAKE_THREAD always.

**Comment by ClarexZhou** [comment]
Run both function and stress tests on below devices. No side effect found.

BYT MinnowBoard | BYT MinnowBoard | APL UP2 | APL UP2 | GLK | CML Chrome | CML_Mantis | ICL_RVP
-- | -- | -- | -- | -- | -- | -- | --
nocodec | I2S RT5682 | nocodec | I2S PCM512x | I2S-DA7219 | I2S RT5682 | HDA ALC3204 | HDA ALC700

File: sound/soc/sof/intel/hda.h
Comment (kv2019i) [inline sound/soc/sof/intel/hda.h:517]: @bardliao I'd have uniform names to this (even if stream irq function is changed in a later patch). E.g. something like:
check_ipc_irq()
check_stream_irq()

**Review by kv2019i — COMMENTED** [review]
Minor note on the naming. Can be changed in a follow-up patch is other issues are cleared.

File: sound/soc/sof/intel/hda.h
Comment (plbossart) [reply sound/soc/sof/intel/hda.h:517]: agree, I'll add an update to fix the naming but we should merge this today since it'll give us one working day to fix things before the week-end.

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by plbossart** [comment]
@fredoh9 can you help here, CI seems to be stuck?

File: sound/soc/sof/intel/hda-stream.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-stream.c:556]: @bardliao @plbossart I find this odd the default return value is true. The logic is correct but typically we always assume its false and return true when conditions are favorable.

**Comment by ranj063** [comment]
@plbossart @bardliao this patch looks OK to merge given that its critical for testing but a follow up with minor changes will be needed.

File: sound/soc/sof/intel/hda-stream.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-stream.c:556]: the initial code assumed an IRQ_WAKE_THREAD as default, so it's acceptable to me.

**Comment by plbossart** [comment]
ok, so no CI today, so will assume things work and move on.

**Pull request merged**
